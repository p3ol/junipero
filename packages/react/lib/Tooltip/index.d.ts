import { Placement } from '@floating-ui/react-dom-interactions';
import {
  Props as clickProps,
} from '@floating-ui/react-dom-interactions/src/hooks/useClick';
import {
  Props as dismissProps,
} from '@floating-ui/react-dom-interactions/src/hooks/useDismiss';
import {
  Props as hoverProps,
} from '@floating-ui/react-dom-interactions/src/hooks/useHover';
import React, { MutableRefObject } from 'react';

export declare type TooltipRef = {
  onpened: Boolean;
  innerRef: MutableRefObject<any>;
  handleRef: MutableRefObject<any>;
  open: () => void;
  close: () => void;
  toggle: () => void;
};

declare interface TooltipProps extends React.ComponentPropsWithRef<any> {
  animate?: (
    tooltipInner: React.ReactNode,
    options: { opened?: Boolean }
  ) => void;
  apparition?: String;
  children?: React.ReactNode;
  className?: String;
  clickOptions?: clickProps;
  container?: React.ReactNode;
  disabled?: Boolean;
  dismissOptions?: dismissProps;
  floatingOptions?: Object;
  hoverOptions?: hoverProps;
  opened?: Boolean;
  text?: String;
  placement?: Placement;
  trigger?: String;
  onToggle?: (props: { opened?: Boolean }) => void;
  ref?: MutableRefObject<TooltipRef | undefined>;
}

declare function Tooltip(props: TooltipProps): JSX.Element;

export default Tooltip;
