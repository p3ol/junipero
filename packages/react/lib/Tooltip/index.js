import {
  Children,
  forwardRef,
  cloneElement,
  useImperativeHandle,
  useReducer,
  useRef,
  useEffect,
} from 'react';
import { createPortal } from 'react-dom';
import { ensureNode, classNames, mockState } from '@junipero/core';
import {
  useFloating,
  useInteractions,
  useClick,
  useHover,
  useDismiss,
  offset,
  flip,
  shift,
  autoUpdate,
} from '@floating-ui/react-dom-interactions';
import PropTypes from 'prop-types';

const Tooltip = forwardRef(({
  animate,
  apparition,
  children,
  className,
  clickOptions,
  container,
  disabled,
  dismissOptions,
  floatingOptions,
  hoverOptions,
  opened,
  text,
  placement = 'top',
  trigger = 'hover',
  onToggle,
  ...rest
}, ref) => {
  const handleRef = useRef();
  const innerRef = useRef();
  const [state, dispatch] = useReducer(mockState, {
    opened: opened ?? false,
  });

  const { x, y, reference, floating, strategy, context } = useFloating({
    open: state.opened,
    onOpenChange: (...args) => onOpenChange(...args),
    placement,
    strategy: 'fixed',
    whileElementsMounted: autoUpdate,
    ...floatingOptions || {},
    middleware: floatingOptions?.middleware || [
      offset(10),
      flip(),
      shift(),
    ],
  });

  const { getReferenceProps, getFloatingProps } = useInteractions([
    useClick(context, {
      enabled: trigger === 'click',
      ...clickOptions || {},
    }),
    useHover(context, {
      enabled: trigger === 'hover',
      ...hoverOptions || {},
    }),
    useDismiss(context, {
      enabled: trigger === 'click',
      ...dismissOptions || {},
    }),
  ]);

  useEffect(() => {
    dispatch({ opened: disabled ? false : !!opened });
  }, [disabled, opened]);

  useImperativeHandle(ref, () => ({
    opened: state.opened,
    innerRef,
    handleRef,
    open,
    close,
    toggle,
    isJunipero: true,
  }));

  const onOpenChange = o => {
    if (disabled && o) {
      return;
    }

    dispatch({ opened: o });
    onToggle?.({ opened: o });
  };

  const toggle = () => {
    if (disabled) {
      return;
    }

    if (state.opened) {
      close();
    } else {
      open();
    }
  };

  const open = () => {
    if (disabled) {
      return;
    }

    dispatch({ opened: true });
    onToggle?.({ opened: true });
  };

  const close = () => {
    if (disabled) {
      return;
    }

    dispatch({ opened: false });
    onToggle?.({ opened: false });
  };

  const setReference = r => {
    handleRef.current = r?.isJunipero ? r.innerRef.current : r;
    reference(r?.isJunipero ? r.innerRef.current : r);
  };

  const setFloatingRef = r => {
    innerRef.current = r;
    floating(r);
  };

  const tooltipInner = (
    <div className="tooltip-inner">
      { text }
    </div>
  );

  const tooltip = (
    <div
      className={classNames(
        'junipero',
        'tooltip',
        className,
      )}
      data-placement={placement}
      style={{
        position: strategy,
        top: y ?? 0,
        left: x ?? 0,
        ...rest.style || {},
      }}
      { ...getFloatingProps() }
      ref={setFloatingRef}
    >
      { animate
        ? animate(tooltipInner, { opened: state.opened })
        : tooltipInner
      }
    </div>
  );

  return (
    <>
      { !children || typeof children === 'string' ? (
        <span
          { ...getReferenceProps() }
          ref={setReference}
        >
          { children }
        </span>
      ) : cloneElement(Children.only(children), {
        ...getReferenceProps(),
        ref: setReference,
      })}

      {
        state.opened || animate || apparition === 'css'
          ? container
            ? createPortal(tooltip, ensureNode(container))
            : tooltip
          : null
      }
    </>
  );
});

Tooltip.displayName = 'Tooltip';
Tooltip.propTypes = {
  animate: PropTypes.func,
  apparition: PropTypes.string,
  container: PropTypes.oneOfType([
    PropTypes.node,
    PropTypes.object,
    PropTypes.func,
    PropTypes.string,
  ]),
  disabled: PropTypes.bool,
  onToggle: PropTypes.func,
  opened: PropTypes.bool,
  placement: PropTypes.string,
  text: PropTypes.oneOfType([
    PropTypes.string,
    PropTypes.object,
    PropTypes.node,
    PropTypes.func,
  ]),
  trigger: PropTypes.string,
  clickOptions: PropTypes.object,
  dismissOptions: PropTypes.object,
  floatingOptions: PropTypes.object,
  hoverOptions: PropTypes.object,
};

export default Tooltip;
