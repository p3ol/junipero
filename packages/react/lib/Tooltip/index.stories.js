import { action } from '@storybook/addon-actions';
import { slideInDownMenu } from '@junipero/transitions';

import Tooltip from './index';

export default { title: 'react/Tooltip' };

export const basic = () => (
  <Tooltip text="Text" onToggle={action('toggle')}>
    Hover me
  </Tooltip>
);

export const click = () => (
  <Tooltip text="Text" onToggle={action('toggle')} trigger="click">
    Click me
  </Tooltip>
);

export const animated = () => (
  <Tooltip text="Text" animate={slideInDownMenu} onToggle={action('toggle')}>
    Hover me
  </Tooltip>
);
