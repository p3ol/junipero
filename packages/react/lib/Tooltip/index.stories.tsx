import { MouseEvent, useRef, useState } from 'react';
import { action } from '@storybook/addon-actions';
import { slideInDownMenu } from '@junipero/transitions';

import Tooltip, { TooltipRef } from './index';

export default { title: 'react/Tooltip' };

export const basic = () => (
  <Tooltip text="Text" onToggle={action('toggle')}>
    Hover me
  </Tooltip>
);

export const click = () => (
  <Tooltip text="Text" onToggle={action('toggle')} trigger="click">
    Click me
  </Tooltip>
);

export const animated = () => (
  <Tooltip text="Text" animate={slideInDownMenu} onToggle={action('toggle')}>
    Hover me
  </Tooltip>
);

export const animatedButStillInDom = () => (
  <Tooltip
    text="Text"
    animate={slideInDownMenu}
    onToggle={action('toggle')}
    apparition="css"
  >
    Hover me
  </Tooltip>
);

export const manuallyUpdate = () => {
  const [pos, setPos] = useState({ x: 0, y: 0 });
  const tooltipRef = useRef<TooltipRef>();

  const handleMouseMove = (e: MouseEvent) => {
    setPos({ x: e.pageX, y: e.pageY });
    tooltipRef.current.update();
  };

  return (
    <div
      style={{ width: 800, height: 800 }}
      onMouseMove={handleMouseMove}
    >
      <Tooltip
        ref={tooltipRef}
        text="Text"
        opened={true}
      >
        <svg
          style={{
            width: 40,
            height: 40,
            borderRadius: '100%',
            background: 'red',
          }}
          transform={`translate(${pos.x}, ${pos.y})`}
        />
      </Tooltip>
    </div>
  );
};
