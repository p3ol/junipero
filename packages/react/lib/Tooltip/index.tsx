import {
  Children,
  forwardRef,
  cloneElement,
  useImperativeHandle,
  useReducer,
  useRef,
  useEffect,
  MutableRefObject,
  ComponentPropsWithRef,
  ReactNode,
} from 'react';
import { createPortal } from 'react-dom';
import { classNames, mockState, omit } from '@junipero/core';
import {
  useFloating,
  useInteractions,
  useClick,
  useHover,
  useDismiss,
  offset,
  flip,
  shift,
  autoUpdate,
  UseDismissProps,
  UseClickProps,
  UseHoverProps,
  Placement,
  UseFloatingOptions,
} from '@floating-ui/react';
import PropTypes from 'prop-types';

import { ForwardedProps, MockState } from '../utils';

export declare type TooltipRef = {
  opened: boolean;
  open(): void;
  close(): void;
  toggle(): void;
  update(): void;
  innerRef: MutableRefObject<any>;
  handleRef: MutableRefObject<any>;
  isJunipero: boolean;
};

export declare interface TooltipProps extends ComponentPropsWithRef<any> {
  apparition?: string;
  children?: ReactNode | JSX.Element;
  className?: string;
  clickOptions?: UseClickProps;
  container?: Element | DocumentFragment;
  disabled?: boolean;
  dismissOptions?: UseDismissProps;
  floatingOptions?: UseFloatingOptions;
  hoverOptions?: UseHoverProps;
  opened?: boolean;
  text?: ReactNode | JSX.Element;
  placement?: Placement;
  trigger?: string;
  animate?(
    tooltipInner: ReactNode | JSX.Element,
    opts?: { opened?: boolean, onExited?: () => void }
  ): JSX.Element | ReactNode;
  onToggle?(props: { opened: boolean }): void;
  ref?: MutableRefObject<TooltipRef | undefined>;
}

const Tooltip = forwardRef(({
  animate,
  apparition,
  children,
  className,
  clickOptions,
  container,
  disabled,
  dismissOptions,
  floatingOptions,
  hoverOptions,
  opened,
  text,
  placement = 'top',
  trigger = 'hover',
  onToggle,
  ...rest
}:TooltipProps, ref) => {
  const handleRef = useRef();
  const innerRef = useRef();

  type TooltipState = {
    opened: boolean,
    visible: boolean
  };

  const [state, dispatch] = useReducer<MockState<TooltipState>>(mockState, {
    opened: opened ?? false,
    visible: opened ?? false,
  });

  const { x, y, refs, strategy, context, update } = useFloating({
    open: state.opened,
    onOpenChange: (...args) => onOpenChange(...args),
    placement,
    strategy: 'fixed',
    whileElementsMounted: autoUpdate,
    ...omit(floatingOptions || {}, ['boundary', 'middleware']),
    middleware: floatingOptions?.middleware || [
      offset(10),
      flip({
        boundary: floatingOptions?.boundary,
      }),
      shift({
        boundary: floatingOptions?.boundary,
      }),
    ],
  });

  const { getReferenceProps, getFloatingProps } = useInteractions([
    useClick(context, {
      enabled: trigger === 'click',
      ...clickOptions || {},
    }),
    useHover(context, {
      enabled: trigger === 'hover',
      ...hoverOptions || {},
    }),
    useDismiss(context, {
      enabled: trigger === 'click',
      ...dismissOptions || {},
    }),
  ]);

  useEffect(() => {
    dispatch({ opened: disabled ? false : !!opened });
  }, [disabled, opened]);

  useImperativeHandle(ref, () => ({
    opened: state.opened,
    innerRef,
    handleRef,
    open,
    close,
    toggle,
    update,
    isJunipero: true,
  }));

  const onOpenChange = (o: boolean, _: Event) => {
    if (disabled && o) {
      return;
    }

    dispatch({ opened: o, visible: true });
    onToggle?.({ opened: o });
  };

  const toggle = () => {
    if (disabled) {
      return;
    }

    if (state.opened) {
      close();
    } else {
      open();
    }
  };

  const open = () => {
    if (disabled) {
      return;
    }

    dispatch({ opened: true, visible: true });
    onToggle?.({ opened: true });
  };

  const close = () => {
    if (disabled) {
      return;
    }

    dispatch({ opened: false });
    onToggle?.({ opened: false });
  };

  const setReference = r => {
    handleRef.current = r?.isJunipero ? r.innerRef.current : r;
    refs.setReference(r?.isJunipero ? r.innerRef.current : r);
  };

  const setFloatingRef = r => {
    innerRef.current = r;
    refs.setFloating(r);
  };

  const onAnimationExit = () => {
    dispatch({ visible: false });
  };

  const tooltipInner = (
    <div className="tooltip-inner">
      { text }
    </div>
  );

  const tooltip = (
    <div
      { ...rest }
      className={classNames(
        'junipero',
        'tooltip',
        className,
      )}
      data-placement={placement}
      style={{
        position: strategy,
        top: y ?? 0,
        left: x ?? 0,
        ...rest.style || {},
      }}
      { ...getFloatingProps() }
      ref={setFloatingRef}
    >
      { animate
        ? animate(tooltipInner, {
          opened: state.opened,
          onExited: onAnimationExit,
        })
        : tooltipInner
      }
    </div>
  );

  return (
    <>
      { !children || typeof children === 'string' ? (
        <span
          { ...getReferenceProps() }
          ref={setReference}
        >
          { children }
        </span>
      ) : cloneElement(Children.only<JSX.Element>(children as JSX.Element), {
        ...getReferenceProps(), //TODO fixme
        ref: setReference,
      }) }

      { state.opened || (animate && state.visible) || apparition === 'css'
        ? container
          ? createPortal(tooltip, container)
          : tooltip
        : null
      }
    </>
  );
}) as ForwardedProps<TooltipProps, TooltipRef>;

Tooltip.displayName = 'Tooltip';
Tooltip.propTypes = {
  animate: PropTypes.func,
  apparition: PropTypes.string,
  container: PropTypes.any,
  disabled: PropTypes.bool,
  onToggle: PropTypes.func,
  opened: PropTypes.bool,
  placement: PropTypes.any,
  text: PropTypes.oneOfType([
    PropTypes.node,
  ]),
  trigger: PropTypes.string,
  clickOptions: PropTypes.object,
  dismissOptions: PropTypes.object,
  floatingOptions: PropTypes.object,
  hoverOptions: PropTypes.object,
};

export default Tooltip;
