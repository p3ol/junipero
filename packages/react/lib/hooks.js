import { useContext } from 'react';

import { FieldControlContext } from './contexts';

export const useFieldControl = () => useContext(FieldControlContext);
