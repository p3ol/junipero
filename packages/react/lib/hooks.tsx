import { useContext } from 'react';

import {
  AlertsContextType,
  AlertsContext,
  DropdownContext,
  DropdownContextType,
  FieldContextType,
  FieldControlContext,
  ListContext,
  ListContextType,
  ToastsContext,
  ToastsContextType,
  ModalContext,
  ModalContextType,
} from './contexts';

export const useFieldControl =
  () => useContext<FieldContextType>(FieldControlContext);
export const useDropdown = () =>
  useContext<DropdownContextType>(DropdownContext);
export const useAlerts = () => useContext<AlertsContextType>(AlertsContext);
export const useList = () => useContext<ListContextType>(ListContext);
export const useToasts = () => useContext<ToastsContextType>(ToastsContext);
export const useModal = () => useContext<ModalContextType>(ModalContext);

export { AlertsContext, DropdownContext, FieldControlContext, ListContext, ToastsContext, ModalContext };
