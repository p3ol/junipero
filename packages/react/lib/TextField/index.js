import {
  forwardRef,
  useEffect,
  useImperativeHandle,
  useReducer,
  useRef,
} from 'react';
import { classNames, mockState, exists } from '@junipero/core';
import PropTypes from 'prop-types';

import { useFieldControl } from '../hooks';

const TextField = forwardRef(({
  min,
  max,
  autoFocus,
  children,
  className,
  valid,
  value,
  disabled = false,
  required = false,
  type = 'text',
  onBlur,
  onChange,
  onFocus,
  onWheel,
  onValidate = (val, { required }) => !!val || !required,
  ...rest
}, ref) => {
  const innerRef = useRef();
  const inputRef = useRef();
  const { update: updateControl } = useFieldControl();
  const [state, dispatch] = useReducer(mockState, {
    value: value ?? '',
    valid: valid ?? false,
    focused: autoFocus ?? false,
    dirty: false,
  });

  useImperativeHandle(ref, () => ({
    innerRef,
    inputRef,
    value: state.value,
    valid: state.valid,
    dirty: state.dirty,
    focused: state.focused,
    focus,
    blur,
    reset,
    setDirty,
    isJunipero: true,
  }));

  useEffect(() => {
    if (exists(value)) {
      state.value = value;
      state.valid = onValidate(state.value, { dirty: state.dirty, required });
      dispatch({ value: state.value, valid: state.valid });
      updateControl?.({ valid: state.valid, dirty: state.dirty });
    }
  }, [value]);

  useEffect(() => {
    dispatch({ valid: valid ?? false });
    updateControl?.({ valid: state.valid });
  }, [valid]);

  const parseValue = value => {
    const value_ = value.replace(',', '.');
    const parsedValue = parseFloat(value_);

    return !isNaN(parsedValue) ? parsedValue : value_;
  };

  const onChange_ = e => {
    /* html input disabled attribute will prevent onChange if present anyway */
    /* istanbul ignore next: just in case */
    if (disabled) {
      return;
    }

    if (type === 'number') {
      value = parseValue(e?.target?.value);
    } else {
      value = e?.target?.value;
    }

    state.value = value;
    state.dirty = true;
    state.valid = onValidate(
      state.value, { dirty: state.dirty, required }
    );
    dispatch({ value: state.value, dirty: state.dirty });
    onChange?.({ value: state.value, valid: state.valid, dirty: state.dirty });
    updateControl?.({ dirty: true, valid: state.valid });
  };

  const onFocus_ = e => {
    dispatch({ focused: true });
    updateControl?.({ focused: true });
    onFocus?.(e);
  };

  const onBlur_ = e => {
    dispatch({ focused: false });
    updateControl?.({ focused: false });
    onBlur?.(e);
  };

  const focus = () => {
    inputRef.current?.focus();
  };

  const blur = () => {
    inputRef.current?.blur();
  };

  const reset = () => {
    dispatch({ value: value ?? '', valid: valid ?? false, dirty: false });
    updateControl?.({ dirty: false, valid: valid ?? false });
  };

  const isEmpty = () =>
    !exists(state.value) || state.value === '';

  const setDirty = dirty => {
    state.dirty = dirty;
    dispatch({ dirty });
    updateControl?.({ dirty });
  };

  const Tag = rest.rows > 1 ? 'textarea' : 'input';

  return (
    <div
      ref={innerRef}
      className={classNames(
        'junipero',
        'text-field',
        state.dirty ? 'dirty' : 'pristine',
        !state.valid && state.dirty ? 'invalid' : 'valid',
        { empty: isEmpty(), focused: state.focused, disabled },
        className
      )}
    >
      <Tag
        { ...rest }
        min={min}
        max={max}
        autoFocus={autoFocus}
        className="field"
        ref={inputRef}
        disabled={disabled}
        required={required}
        type={Tag === 'textarea' ? undefined : type}
        value={state.value}
        onChange={onChange_}
        onFocus={onFocus_}
        onBlur={onBlur_}
        onWheel={onWheel ?? (() => inputRef.current?.blur())}
      />
      { children }
    </div>
  );
});

TextField.displayName = 'TextField';
TextField.propTypes = {
  min: PropTypes.number,
  max: PropTypes.number,
  autoFocus: PropTypes.bool,
  disabled: PropTypes.bool,
  required: PropTypes.bool,
  type: PropTypes.string,
  valid: PropTypes.bool,
  value: PropTypes.oneOfType([
    PropTypes.string,
    PropTypes.number,
  ]),
  onBlur: PropTypes.func,
  onChange: PropTypes.func,
  onFocus: PropTypes.func,
  onWheel: PropTypes.func,
  onValidate: PropTypes.func,
};

export default TextField;
