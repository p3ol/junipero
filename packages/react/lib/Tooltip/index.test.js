import { act, render, fireEvent } from '@testing-library/react';

import Tooltip from './';

describe('<Tooltip />', () => {
  it('should render', () => {
    const { container, unmount } = render(
      <Tooltip text="Bar"><span>Foo</span></Tooltip>
    );

    expect(container).toMatchSnapshot();
    fireEvent.mouseEnter(container.querySelector('span'));
    expect(container).toMatchSnapshot();
    fireEvent.mouseLeave(container.querySelector('span'));
    expect(container).toMatchSnapshot();

    unmount();
  });

  it('should also render tooltip when using click as trigger', async () => {
    const { container, unmount } = render(
      <Tooltip text="Bar" trigger="click">Foo</Tooltip>
    );

    expect(container).toMatchSnapshot();
    fireEvent.click(container.querySelector('span'));
    expect(container).toMatchSnapshot();
    fireEvent.click(container.querySelector('span'));
    expect(container).toMatchSnapshot();

    unmount();
  });

  it('should automatically close tooltip when disabled prop changes', () => {
    const { container, rerender, unmount } = render(
      <Tooltip text="Bar">Foo</Tooltip>
    );
    fireEvent.mouseEnter(container.querySelector('span'));
    expect(container).toMatchSnapshot();
    rerender(<Tooltip disabled text="Bar">Foo</Tooltip>);
    expect(container).toMatchSnapshot();
    unmount();
  });

  it('shouldn\'t open tooltip if disabled', () => {
    const { container, unmount } = render(
      <Tooltip text="Bar" disabled>
        Foo
      </Tooltip>
    );
    fireEvent.mouseEnter(container.querySelector('span'));
    expect(container).toMatchSnapshot();
    unmount();
  });

  it('should be able render inside another container using portals', () => {
    const { container: target, unmount: unmountContainer } = render(
      <div className="test-container" />
    );

    const { container, unmount } = render(
      <Tooltip text="Bar" container=".test-container">Foo</Tooltip>
    );
    fireEvent.mouseEnter(container.querySelector('span'));
    expect(target).toMatchSnapshot();
    unmountContainer();
    unmount();
  });
});
