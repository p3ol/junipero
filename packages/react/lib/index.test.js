import { useReducer } from 'react';
import { render, screen, fireEvent, waitFor } from '@testing-library/react';
import { mockState } from '@junipero/core';
import { useTimeout } from '@junipero/hooks';

import TextField from './TextField';

describe('mockState: inside react', () => {
  it('should fail update the state correctly and return the old ' +
    'state', async () => {
    const Comp = () => {
      const [state, dispatch] = useReducer(mockState, {
        value: {
          text: 'This is a test',
          content: [],
        },
      });

      useTimeout(() => {
        dispatch({ value: { ...state.value, content: ['foo'] } });
      }, 1, []);

      const onTextChange = ({ value }) => {
        dispatch({ value: { ...state.value, text: value } });
      };

      const onContentChange = () => {
        dispatch({
          value: { ...state.value, content: [...state.value.content, 'bar'] },
        });
      };

      return (
        <div>
          <TextField
            data-testid="Field"
            value={state.value.text}
            onChange={onTextChange}
          />
          <button onClick={onContentChange}>Add content</button>
          <div>Text: { state.value.text }</div>
          <div>Content: { state.value.content.join(', ') }</div>
        </div>
      );
    };

    const { container, unmount } = render(<Comp />);

    fireEvent.change(screen.getByTestId('Field'),
      { target: { value: 'This is a test 2' } });

    await waitFor(() => screen.getByText('Content: foo'));
    expect(container).toMatchSnapshot('With updated text field value');

    fireEvent.click(screen.getByText('Add content'));

    expect(container)
      .toMatchSnapshot('With updated content but with old text value');

    unmount();
  });

  it('should not fail update the state correctly and return the new ' +
    'state when using mockState with a callback', async () => {
    const Comp = () => {
      const [state, dispatch] = useReducer(mockState, {
        value: {
          text: 'This is a test',
          content: [],
        },
      });

      useTimeout(() => {
        dispatch(s => ({ value: { ...s.value, content: ['foo'] } }));
      }, 1, []);

      const onTextChange = ({ value }) => {
        dispatch(s => ({ value: { ...s.value, text: value } }));
      };

      const onContentChange = () => {
        dispatch(s => ({
          value: { ...s.value, content: [...s.value.content, 'bar'] },
        }));
      };

      return (
        <div>
          <TextField
            data-testid="Field"
            value={state.value.text}
            onChange={onTextChange}
          />
          <button onClick={onContentChange}>Add content</button>
          <div>Text: { state.value.text }</div>
          <div>Content: { state.value.content.join(', ') }</div>
        </div>
      );
    };

    const { container, unmount } = render(<Comp />);

    fireEvent.change(screen.getByTestId('Field'),
      { target: { value: 'This is a test 2' } });

    await waitFor(() => screen.getByText('Content: foo'));
    expect(container).toMatchSnapshot('With updated text field value');

    fireEvent.click(screen.getByText('Add content'));

    expect(container)
      .toMatchSnapshot('With updated content AND updated text field valud');

    unmount();
  });
});
