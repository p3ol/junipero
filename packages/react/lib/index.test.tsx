import { useReducer } from 'react';
import { render, screen, fireEvent, waitFor } from '@testing-library/react';
import { mockState } from '@junipero/core';
import { useTimeout } from '@junipero/hooks';

import type { FieldContent } from './types';
import TextField from './TextField';

interface State {
  value: {
    text: string;
    content: string[];
  };
}

describe('mockState: inside react', () => {
  it('should correctly update the state and return the new ' +
    'state when using mockState with a callback', async () => {
    const Comp = () => {
      const [state, dispatch] = useReducer(mockState<State>, {
        value: {
          text: 'This is a test',
          content: [],
        },
      });

      useTimeout(() => {
        dispatch(s => ({ value: { ...s.value, content: ['foo'] } }));
      }, 1, []);

      const onTextChange = ({ value }: FieldContent) => {
        dispatch(s => ({ value: { ...s.value, text: value } }));
      };

      const onContentChange = () => {
        dispatch(s => ({
          value: { ...s.value, content: [...s.value.content, 'bar'] },
        }));
      };

      return (
        <div>
          <TextField
            data-testid="Field"
            value={state.value.text}
            onChange={onTextChange}
          />
          <button onClick={onContentChange}>Add content</button>
          <div>Text: { state.value.text }</div>
          <div>Content: { state.value.content.join(', ') }</div>
        </div>
      );
    };

    const { container, unmount } = render(<Comp />);

    fireEvent.change(screen.getByTestId('Field'),
      { target: { value: 'This is a test 2' } });

    await waitFor(() => screen.getByText('Content: foo'));
    expect(container).toMatchSnapshot('With updated text field value');

    fireEvent.click(screen.getByText('Add content'));

    expect(container)
      .toMatchSnapshot('With updated content AND updated text field valud');

    unmount();
  });
});
