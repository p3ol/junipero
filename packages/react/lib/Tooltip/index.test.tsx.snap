// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Tooltip /> should allow to animate tooltip: Enter: Tooltip in dom 1`] = `
<div>
  <span>
    Foo
  </span>
  <div
    class="junipero tooltip"
    data-floating-ui-focusable=""
    data-placement="top"
    style="position: fixed; top: 0px; left: 0px;"
    tabindex="-1"
  >
    <div
      class="tooltip-inner jp-slide-in-down-menu-enter-done"
    >
      Bar
    </div>
  </div>
</div>
`;

exports[`<Tooltip /> should allow to animate tooltip: Enter: Tooltip not in dom 1`] = `
<div>
  <span>
    Foo
  </span>
</div>
`;

exports[`<Tooltip /> should allow to animate tooltip: Exit: Tooltip in dom 1`] = `
<div>
  <span>
    Foo
  </span>
  <div
    class="junipero tooltip"
    data-floating-ui-focusable=""
    data-placement="top"
    style="position: fixed; top: 0px; left: 0px;"
    tabindex="-1"
  >
    <div
      class="tooltip-inner jp-slide-in-down-menu-exit"
    >
      Bar
    </div>
  </div>
</div>
`;

exports[`<Tooltip /> should allow to animate tooltip: Exit: Tooltip not in dom 1`] = `
<div>
  <span>
    Foo
  </span>
</div>
`;

exports[`<Tooltip /> should also render tooltip when using click as trigger 1`] = `
<div>
  <span>
    Foo
  </span>
</div>
`;

exports[`<Tooltip /> should also render tooltip when using click as trigger 2`] = `
<div>
  <span>
    Foo
  </span>
  <div
    class="junipero tooltip"
    data-floating-ui-focusable=""
    data-placement="top"
    style="position: fixed; top: 0px; left: 0px;"
    tabindex="-1"
  >
    <div
      class="tooltip-inner"
    >
      Bar
    </div>
  </div>
</div>
`;

exports[`<Tooltip /> should also render tooltip when using click as trigger 3`] = `
<div>
  <span>
    Foo
  </span>
</div>
`;

exports[`<Tooltip /> should automatically close tooltip when disabled prop changes 1`] = `
<div>
  <span>
    Foo
  </span>
  <div
    class="junipero tooltip"
    data-floating-ui-focusable=""
    data-placement="top"
    style="position: fixed; top: 0px; left: 0px;"
    tabindex="-1"
  >
    <div
      class="tooltip-inner"
    >
      Bar
    </div>
  </div>
</div>
`;

exports[`<Tooltip /> should automatically close tooltip when disabled prop changes 2`] = `
<div>
  <span>
    Foo
  </span>
</div>
`;

exports[`<Tooltip /> should be able render inside another container using portals 1`] = `
<div>
  <div
    class="test-container"
  >
    <div
      class="junipero tooltip"
      data-floating-ui-focusable=""
      data-placement="top"
      style="position: fixed; top: 0px; left: 0px;"
      tabindex="-1"
    >
      <div
        class="tooltip-inner"
      >
        Bar
      </div>
    </div>
  </div>
</div>
`;

exports[`<Tooltip /> should keep animated tooltip in dom if apparition is css: Enter: Tooltip already in dom 1`] = `
<div>
  <span>
    Foo
  </span>
  <div
    class="junipero tooltip"
    data-floating-ui-focusable=""
    data-placement="top"
    style="position: fixed; top: 0px; left: 0px;"
    tabindex="-1"
  />
</div>
`;

exports[`<Tooltip /> should keep animated tooltip in dom if apparition is css: Enter: Tooltip still in dom 1`] = `
<div>
  <span>
    Foo
  </span>
  <div
    class="junipero tooltip"
    data-floating-ui-focusable=""
    data-placement="top"
    style="position: fixed; top: 0px; left: 0px;"
    tabindex="-1"
  >
    <div
      class="tooltip-inner jp-slide-in-down-menu-enter-done"
    >
      Bar
    </div>
  </div>
</div>
`;

exports[`<Tooltip /> should keep animated tooltip in dom if apparition is css: Exit: Tooltip still in dom 1`] = `
<div>
  <span>
    Foo
  </span>
  <div
    class="junipero tooltip"
    data-floating-ui-focusable=""
    data-placement="top"
    style="position: fixed; top: 0px; left: 0px;"
    tabindex="-1"
  />
</div>
`;

exports[`<Tooltip /> should render 1`] = `
<div>
  <span>
    Foo
  </span>
</div>
`;

exports[`<Tooltip /> should render 2`] = `
<div>
  <span>
    Foo
  </span>
  <div
    class="junipero tooltip"
    data-floating-ui-focusable=""
    data-placement="top"
    style="position: fixed; top: 0px; left: 0px;"
    tabindex="-1"
  >
    <div
      class="tooltip-inner"
    >
      Bar
    </div>
  </div>
</div>
`;

exports[`<Tooltip /> should render 3`] = `
<div>
  <span>
    Foo
  </span>
</div>
`;

exports[`<Tooltip /> shouldn't open tooltip if disabled 1`] = `
<div>
  <span>
    Foo
  </span>
</div>
`;
